File: QuasarCode_Library/Games/Spinners/ISpinner.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace QuasarCode.Library.Games.Spinners
{
    /// <summary>
    /// Interface for all Spinner objects.
    /// </summary>
    /// <typeparam name="T">Type of result data held by the spinner.</typeparam>
    public interface ISpinner<T>
    {
        /// <summary>
        /// Array of side labels
        /// </summary>
        T[] Labels { get; }

        /// <summary>
        /// Number of sides/labels
        /// </summary>
        int Sides { get; }

        /// <summary>
        /// Randomly selects a side and retuns the side and the corisponding label.
        /// </summary>
        /// <returns>Tuple with int and label type</returns>
        Tuple<int, T> ContextSpin();

        /// <summary>
        /// Randomly selects a side and retuns the corisponding label.
        /// </summary>
        /// <returns>Side label</returns>
        T Spin();

        /// <summary>
        /// Randomly selects a side and retuns the side number.
        /// </summary>
        /// <returns>Side number</returns>
        int IndexSpin();
    }
}

File: QuasarCode_Python/QuasarCode/Games/Spinners/_ISpinner.py
class ISpinner(object):
    """
    Interface for all Spinner objects.
    """

    @property
    def Labels(self) -> list:
        """
        Array of side labels
        """
        raise NotImplementedError()

    @property
    def Sides(self) -> int:
        """
        Number of sides/labels
        """
        raise NotImplementedError()

    def ContextSpin(self):
        """
        Randomly selects a side and retuns the side and the corisponding label.
        """
        raise NotImplementedError()

    def Spin(self):
        """
        Randomly selects a side and retuns the corisponding label.
        """
        raise NotImplementedError()

    def IndexSpin(self):
        """
        Randomly selects a side and retuns the side number.
        """
        raise NotImplementedError()